File: src/main/java/net/casualuhc/uhcmod/utils/screen/RuleScreen.java
package net.casualuhc.uhcmod.utils.screen;

import net.casualuhc.uhcmod.utils.gamesettings.GameSetting;
import net.casualuhc.uhcmod.utils.gamesettings.GameSettings;
import net.casualuhc.uhcmod.utils.uhc.ItemUtils;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.screen.SimpleNamedScreenHandlerFactory;
import net.minecraft.screen.slot.SlotActionType;
import net.minecraft.text.Text;

import java.util.List;

public class RuleScreen extends CustomScreen {
	private final int page;

	public RuleScreen(PlayerInventory playerInventory, int syncId, int page) {
		super(playerInventory, syncId, 6);
		this.page = page;

		this.modifyInventory(inventory -> {
			List<GameSetting<?>> gameSettings = GameSettings.RULES.values().stream().skip(5L * this.page).toList();

			for (int i = 1; i <= 5; i++) {
				if (gameSettings.size() <= (i - 1)) {
					break;
				}
				int startSlot = i * 9;
				GameSetting<?> gameSetting = gameSettings.get(i - 1);
				inventory.setStack(startSlot, gameSetting.getDisplay());

				startSlot += 2;
				List<ItemStack> itemStacks = gameSetting.getOptions().keySet().stream().limit(7).toList();
				int options = itemStacks.size();
				for (int j = 0; j < options; j++) {
					inventory.setStack(startSlot + j, itemStacks.get(j));
				}
			}

			for (int i = 1; i < 8; i++) {
				inventory.setStack(i, Items.GRAY_STAINED_GLASS.getDefaultStack().setCustomName(Text.literal("")));
			}

			inventory.setStack(0, ItemUtils.literalNamed(Items.RED_STAINED_GLASS, "Previous"));
			inventory.setStack(8, ItemUtils.literalNamed(Items.GREEN_STAINED_GLASS, "Next"));
		});
	}

	@Override
	public void onSlotClick(int slotIndex, int button, SlotActionType actionType, PlayerEntity player) {
		if (slotIndex == 0 && this.page > 0) {
			player.openHandledScreen(createScreenFactory(this.page - 1));
			return;
		}
		if (slotIndex == 8) {
			player.openHandledScreen(createScreenFactory(this.page + 1));
			return;
		}
		if (slotIndex < 9 || slotIndex > 53 || slotIndex % 9 == 0) {
			return;
		}
		int row = Math.floorDiv(slotIndex, 9);
		ItemStack ruleStack = this.slots.get(row * 9).getStack();
		ItemStack clickedStack = this.slots.get(slotIndex).getStack();

		if (clickedStack.isEmpty()) {
			return;
		}

		GameSetting<?> setting = GameSettings.RULES.get(ruleStack);
		if (setting != null) {
			setting.setValueFromOption(clickedStack);
			Text message = ruleStack.getName().copy().append(" has been set to ").append(clickedStack.getName());
			player.sendMessage(message, false);
		}
	}

	public static SimpleNamedScreenHandlerFactory createScreenFactory(int page) {
		if (page < 0) {
			return null;
		}
		return new SimpleNamedScreenHandlerFactory((syncId, inv, player) -> {
			return new RuleScreen(inv, syncId, page);
		}, Text.literal("Rule Config Screen"));
	}
}

File: src/main/kotlin/net/casualuhc/uhcmod/screen/RuleScreen.kt
package net.casualuhc.uhcmod.screen

import net.casualuhc.arcade.utils.ItemUtils.literalNamed
import net.casualuhc.uhcmod.settings.GameSettings
import net.minecraft.network.chat.Component
import net.minecraft.world.SimpleMenuProvider
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.ClickType
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items

class RuleScreen(inventory: Inventory, syncId: Int, private val page: Int): CustomScreen(inventory, syncId, 6) {
    init {
        this.modifyInventory {
            val gameSettings = GameSettings.RULES.values.stream().skip(5L * page).toList()
            for (i in 1..5) {
                if (gameSettings.size <= i - 1) {
                    break
                }
                var startSlot = i * 9
                val gameSetting = gameSettings[i - 1]
                it.setItem(startSlot, gameSetting.display)

                startSlot += 2
                val itemStacks = gameSetting.options.keys.stream().limit(7).toList()
                for (j in 0 until itemStacks.size) {
                    it.setItem(startSlot + j, itemStacks[j])
                }
            }

            for (i in 1..7) {
                it.setItem(i, Items.GRAY_STAINED_GLASS.defaultInstance.setHoverName(Component.empty()))
            }

            it.setItem(0, Items.RED_STAINED_GLASS.literalNamed("Previous"))
            it.setItem(8, Items.GREEN_STAINED_GLASS.literalNamed("Next"))
        }
    }

    override fun clicked(slotId: Int, button: Int, clickType: ClickType, player: Player) {
        if (slotId == 0 && this.page > 0) {
            player.openMenu(createScreenFactory(this.page - 1))
            return
        }
        if (slotId == 8) {
            player.openMenu(createScreenFactory(this.page + 1))
            return
        }
        if (slotId < 9 || slotId > 53 || slotId % 9 == 0) {
            return
        }
        val row = Math.floorDiv(slotId, 9)
        val ruleStack: ItemStack = this.slots[row * 9].item
        val clickedStack: ItemStack = this.slots[slotId].item

        if (clickedStack.isEmpty) {
            return
        }

        val setting = GameSettings.RULES[ruleStack]
        if (setting != null) {
            setting.setValueFromOption(clickedStack)
            val message = Component.literal("${setting.name} has been set to ").append(clickedStack.displayName)
            player.sendSystemMessage(message)
        }
    }

    companion object {
        fun createScreenFactory(page: Int): SimpleMenuProvider? {
            return if (page < 0) null else SimpleMenuProvider(
                { syncId, inv, _ -> RuleScreen(inv, syncId, page) },
                Component.literal("Rule Config Screen")
            )
        }
    }
}
